File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveV8OptionsName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("v8.d.ts")) return null

    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    let parameter = getParent(node)
    if (ts.isParenthesizedTypeNode(parameter)) {
        parameter = getParent(parameter)
    }
    if (ts.isUnionTypeNode(parameter)) {
        parameter = getParent(parameter)
    }

    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null

    const parameterName = parameter.name.text

    let parentName

    const signatureDeclaration = getParent(parameter)
    if (!signatureDeclaration) return null

    if (ts.isFunctionDeclaration(signatureDeclaration)) {
        if (signatureDeclaration.name === undefined) return null

        parentName = signatureDeclaration.name.text
    } else if (ts.isCallSignatureDeclaration(signatureDeclaration)) {
        const interfaceNode = getParent(signatureDeclaration)
        if (!interfaceNode) return null
        if (!ts.isInterfaceDeclaration(interfaceNode)) return null

        parentName = interfaceNode.name.text
    } else if (ts.isMethodSignature(signatureDeclaration)) {
        if (signatureDeclaration.name === undefined) return null

        const interfaceNode = getParent(signatureDeclaration)
        if (!interfaceNode) return null
        if (!ts.isInterfaceDeclaration(interfaceNode)) return null

        parentName = `${karakum.capitalize(interfaceNode.name.text)}${karakum.capitalize(signatureDeclaration.name.text)}`
    } else {
        return null
    }

    if (
        parentName === "queryObjects"
    ) {
        let infix = ""

        if (
            ts.isTypeLiteralNode(node)
            && node.members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "format"
                && ts.isLiteralTypeNode(member.type)
                && ts.isStringLiteral(member.type.literal)
                && member.type.literal.text === "count"
            ))
        ) {
            infix = "Count"
        }

        if (
            ts.isTypeLiteralNode(node)
            && node.members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "format"
                && ts.isLiteralTypeNode(member.type)
                && ts.isStringLiteral(member.type.literal)
                && member.type.literal.text === "summary"
            ))
        ) {
            infix = "Summary"
        }

        return `${karakum.capitalize(parentName)}${infix}${karakum.capitalize(parameterName)}`
    }

    return `${karakum.capitalize(parentName)}${karakum.capitalize(parameterName)}`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveV8OptionsName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import node.karakum.util.nullable
import typescript.Node
import typescript.asArray
import typescript.isCallSignatureDeclaration
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isLiteralTypeNode
import typescript.isMethodSignature
import typescript.isParameter
import typescript.isParenthesizedTypeNode
import typescript.isPropertySignature
import typescript.isStringLiteral
import typescript.isTypeLiteralNode
import typescript.isUnionTypeNode

fun resolveV8OptionsName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    ensure(sourceFileName.endsWith("v8.d.ts"))

    // exclude literal types
    ensure(!isLiteralTypeNode(node))

    val parameter = ensureNotNull(
        typeScriptService.getParent(node)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
            ?.let {
                if (isUnionTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text

    val signatureDeclaration = ensureNotNull(typeScriptService.getParent(parameter))
    val parentName = ensureNotNull(
        nullable {
            ensure(isFunctionDeclaration(signatureDeclaration))
            signatureDeclaration.name?.text
        } ?: nullable {
            ensure(isCallSignatureDeclaration(signatureDeclaration))

            val interfaceNode = ensureNotNull(typeScriptService.getParent(signatureDeclaration))
            ensure(isInterfaceDeclaration((interfaceNode)))

            interfaceNode.name.text
        } ?: nullable {
            ensure(isMethodSignature(signatureDeclaration))

            val methodNameNode = signatureDeclaration.name
            ensure(isIdentifier(methodNameNode))

            val interfaceNode = ensureNotNull(typeScriptService.getParent(signatureDeclaration))
            ensure(isInterfaceDeclaration((interfaceNode)))

            "${interfaceNode.name.text.replaceFirstChar { it.titlecase() }}${methodNameNode.text.replaceFirstChar { it.titlecase() }}"
        }
    )

    nullable {
        ensure(parentName == "queryObjects")

        nullable {
            ensure(isTypeLiteralNode(node))
            ensure(node.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "format")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))

                    val propertyTypeLiteral = propertyType.literal
                    ensure(isStringLiteral(propertyTypeLiteral))
                    ensure(propertyTypeLiteral.text == "count")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}Count${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(isTypeLiteralNode(node))
            ensure(node.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "format")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))

                    val propertyTypeLiteral = propertyType.literal
                    ensure(isStringLiteral(propertyTypeLiteral))
                    ensure(propertyTypeLiteral.text == "summary")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}Summary${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: run {
        "${parentName.replaceFirstChar { it.titlecase() }}${parameterName.replaceFirstChar { it.titlecase() }}"
    }
}
